Add support to disable domain on schedule

iliajie opened the pull request:
Hello Jamie!

There have been a few requests in the past for the ability to disable a virtual server on a simple schedule. I have finally decided to implement this feature. The old Virtualmin Disable module is obsolete and overkill, requiring the installation of the `at` command. This PR implements the feature in a straightforward and simple way.

Please have a look:

<img width="1414" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/6de6c4ac-ec4d-46f6-88c5-3dcc85259b08">


![image](https://github.com/virtualmin/virtualmin-gpl/assets/4426533/08116844-13e6-4397-8daa-7780b0810b4b)

jcameron commented:
This is an intersting feature, but I think it shouldn't go on the Edit Virtual Server page. Instead it should be either an extra menu item under Disable and Delete, or a new button on the Disable page..

iliajie commented:
> This is an intersting feature, but I think it shouldn't go on the Edit Virtual Server page. Instead it should be either an extra menu item under Disable and Delete, or a new button on the Disable page..

Well, the point was not to make it a separate link in the menu as it would involve a lot of additional complexity with access and limits.

Having it on the Disable Domain page makes sense, but it also makes it easy to forget when domains are created for development purposes, forcing to setup disable schedule with an additional step.

However, as I think about it now, it would also make sense to add an extra config option to set it automatically at server creation time. Then, having it factored out to the Disable Domain page would be easier, and won't require taking an additional step.

But still, what if a master admin wants to have different domains configured differently? Either way, it seems to add an additional step.

So, maybe having it under the Edit Virtual Server page is fine? It's a tiny and compact feature that won't add any complexity?

jcameron commented:
No I really disagree strongly that it should go on Edit Virtual Server. We already have a page specifically for disabling domains, and so any new feature to disable on schedule should go there.

iliajie commented:
Alright, I will move it to Disable Domain page..

iliajie commented:
This time, I approached the problem with proper design considerations in mind.

I moved the functionality to the **Disable Virtual Server** page and added an option to the templates. If an admin wants a domain to be automatically disabled after creation in a certain amount of time, they just need to create a template with this option enabled.

Here are some screenshots for clarity:

<img width="1125" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/632747b0-b21e-471c-850b-bd113f89fbb7">

<img width="1127" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/7de753c1-397c-4cfe-bf13-425053bfaf78">

<img width="1127" alt="image" src="https://github.com/virtualmin/virtualmin-gpl/assets/4426533/744e97d1-89ee-4e56-bff6-183092424e62">

Jamie, please review!

marclaporte commented:
> If an admin wants a domain to be automatically disabled after creation in a certain amount of time, they just need to create a template with this option enabled

This is great!!

jcameron commented:
> If an admin wants a domain to be automatically disabled after creation in a certain amount of time, they just need to create a template with this option enabled

Fun fact - this kind of feature was responsible for this outage https://arstechnica.com/gadgets/2024/05/google-cloud-accidentally-nukes-customer-account-causes-two-weeks-of-downtime/

jcameron commented on disable-domain.pl:
```diff
@@ -75,29 +86,72 @@ package virtual_server;
-		if (!$sd->{'disabled'}) {
-			push(@doms, $sd);
+if ($scheduled) {
+	if ($schedule eq "show") {
```
There's no need for this functionality here, instead include the scheduled disable time in `list-domains` output

iliajie replied:
> There's no need for this functionality here, instead include the scheduled disable time in `list-domains` output

Alright, that was removed and I have already added it to `list-domains`.

marclaporte commented:
> this kind of feature was responsible for this outage

We plan to stick to only disabling sites automatically. Any deletions will require a manual review.

marclaporte commented:
And I trust that Virtualmin remote backups are write-only? So even if we delete a Virtualmin Virtual Server, the backups persist on AWS, etc.?

jcameron commented on virtual-server-lib-funcs.pl:
```diff
@@ -8924,6 +8930,32 @@ sub call_feature_delete
+		local $main::error_must_die = 1;
+		my $disabled_auto = &make_date($d->{'disabled_auto'});
+		&disable_virtual_server($d, 'schedule',
+			&text('disable_autodisabledone', $disabled_auto));
```
Shouldn't we clear disabled_auto once the disable is done?

iliajie replied:
Yes, and we do, in `disable_virtual_server` and `enable_virtual_server` since those save the domain.

jcameron replied:
Oh cool, thanks!

jcameron commented on virtual-server-lib-funcs.pl:
```diff
@@ -8924,6 +8930,32 @@ sub call_feature_delete
+{
+foreach my $d (grep {
+		!$_->{'disabled'} && $_->{'disabled_auto'} &&
+		$_->{'disabled_auto'} =~ /(\d{10,})/ && 
```
Why would this ever be set to something that's not a number?

iliajie replied:
This is a fail-safe measure to ensure that the value is a timestamp and nothing else, in case it was manually set to something else by the user.

jcameron replied:
Where could the user set it to something else though? Presumably we check for a valid input in `disable-domain.pl`

I don't think there's a lot of value in doing these kinds of internal checks of Virtualmin's own data, because if we can't trust it then there are hundreds of places it could go wrong.

iliajie replied:
> Where could the user set it to something else though? Presumably we check for a valid input in `disable-domain.pl`
> 
> I don't think there's a lot of value in doing these kinds of internal checks of Virtualmin's own data, because if we can't trust it then there are hundreds of places it could go wrong.

It would only be possible if a user sets it manually by editing domain config file. I don't see how it could be done otherwise (by Virtualmin).

Code like `$_->{'disabled_auto'} =~ /(\d{10,})/` emphasizes that this string is expected to be a timestamp. I don't think it's bad in any way; it just makes the code easier to understand.


> I don't think there's a lot of value in doing these kinds of internal checks of Virtualmin's own data, because if we can't trust it then there are hundreds of places it could go wrong.

Agreed!

jcameron replied:
I mean it's not bad, but I find that unnecessary code is confusing because it makes future readers wonder why this check is needed..

iliajie replied:
> I mean it's not bad, but I find that unnecessary code

This code isn't unnecessary _potentially_  speaking, as it could minimize the chance of having a bug, as mentioned in the Google article you linked earlier.  :--)

> it makes future readers wonder why this check is needed..

I understand your point &mdash; this code isn't there to cover some unexpected behavior on Virtualmin's side if this is your initial concern. On the contrary, having a check for a number of 10 or more digits explicitly clarifies what this code does.

jcameron replied:
Seriously, I think it's best to just remove it. If I was a new programmer on the project and saw that code, I would tear my hair out wondering situation what this check is validating for.

iliajie replied:
> If I was a new programmer on the project and saw that code, I would tear my hair out

LOL! They would have torn their hair out anyway!



> Seriously, I think it's best to just remove it. We certainly can't protect against all possible random edits to domains files!

I could, but I'd rather not. What if something goes wrong and the system's date and time resets to January 1, 1970, 12:00:00 &mdash; it's always best to do sanity check, in particular when we're going to list and then disable all domains.

A new user will Google it. This isn't the most complicated code around the project ..

Is this code really that big of a problem that we're spending so much time on it?

jcameron replied:
Sure this line isn't that important on it's own, but I'm trying to set general principles for the overall Webmin codebase to keep it clean and maintainable.  The point I'm trying to make is that every line of code is a *liability* not an asset, and so we should never write code that is unnecessary. That means that checks for impossible conditions shouldn't be added unless the outcome is catastrophic..

iliajie replied:
> Sure this line isn't that important on it's own, but I'm trying to set general principles for the overall Webmin codebase to keep it clean and maintainable.

This is very well understood, and I share the same principles!



> Sure this line isn't that important on it's own, but I'm trying to set general principles for the overall Webmin codebase to keep it clean and maintainable. The point I'm trying to make is that every line of code is a _liability_ not an asset, and so we should never write code that is unnecessary.

Yes, I perfectly agree!

> That means that checks for impossible conditions shouldn't be added unless the outcome is catastrophic..

This particular case represents a catastrophic outcome if things go wrong. I don't want to leave any chance for this to happen, which is why I added this check in the first place. As we don't use warnings for this library and don't need support for timestamps after November 20, 2286 (what if we do?), I have simplified the code a bit. However, there's no guarantee that we will make it past January 19, 2038 anyways ..  :--)

jcameron replied:
What if we just checked that `disabled_auto` was non-empty? That would cover by far the most common failure case of the field not being set..

iliajie replied:
> What if we just checked that `disabled_auto` was non-empty?

As I previously explained, that could be potentially fragile. I don't want it. I want the check to be explicitly performed for a timestamp.

jcameron replied:
Can we at least extract this out into a function called `is_timestamp` then, which checks if it's a reasonable number? That will be easier to read.

iliajie replied:
> Can we at least extract this out into a function called `is_timestamp` then, which checks if it's a reasonable number? That will be easier to read.

Yes, this is an excellent idea! Implemented here https://github.com/virtualmin/virtualmin-gpl/pull/832/commits/a17bbb9634ce1f12609b2ca2cb82c6b68197ecfc.

Thanks!

jcameron commented on disable_domain.cgi:
```diff
@@ -8,88 +8,180 @@ $d = &get_domain($in{'dom'});
 
-&ui_print_footer(&domain_footer_link($d),
-	"", $text{'index_return'});
+		# Details of current cert
```
cert?

iliajie replied:
It accidentally ended up here as I copied and pasted the tabs API from the `cert_form.cgi` file.

jcameron commented:
> And I trust that Virtualmin remote backups are write-only? So even if we delete a Virtualmin Virtual Server, the backups persist on AWS, etc.?

@marclaporte yes, deleting a domain does NOT delete it's backups! Then would only be removed if you setup purging of old backups, or over-write an existing one.
